dpe/src/commands/sign.rs

```diff
@@ -46,12 +51,26 @@ impl<C: Crypto> CommandExecution<C> for SignCmd {
             return Err(DpeErrorCode::InvalidArgument);
         }
 
-        let cdi = dpe.derive_cdi(idx)?;
         let digest = Digest::new(&self.digest, DPE_PROFILE.alg_len())
             .map_err(|_| DpeErrorCode::InternalError)?;
-        let EcdsaSig { r, s } =
-            C::ecdsa_sign_with_derived(DPE_PROFILE.alg_len(), &cdi, &self.label, b"ECC", &digest)
-                .map_err(|_| DpeErrorCode::InternalError)?;
+        let algs = DPE_PROFILE.alg_len();
+        let priv_key = if self.uses_nd_derivation() {
+            dpe.contexts[idx].cached_priv_key.take().unwrap_or({
+                let cdi = dpe.derive_cdi(idx, true)?;
+                C::derive_ecdsa_key(algs, &cdi, &self.label, b"ECC")
+            })
+        } else {
+            let cdi = dpe.derive_cdi(idx, false)?;
+            C::derive_ecdsa_key(algs, &cdi, &self.label, b"ECC")
+        };
+
+        let EcdsaSig { r, s } = C::ecdsa_sign_with_derived(algs, &digest, &priv_key)
+            .map_err(|_| DpeErrorCode::InternalError)?;
+        // cache private key
+        if self.uses_nd_derivation() {
+            let context = &mut dpe.contexts[idx];
+            context.cached_priv_key.replace(priv_key);
```

Comment:
nit: Not sure introducing a variable helps here. I would expect the same usage as line 58.

```rust
dpe.contexts[idx].cached_priv_key.replace(priv_key);
```